src/almanack/processing/generate_data.py

```diff
@@ -0,0 +1,151 @@
+            file_names,
+        )
+        # Calculate the time span between commits in days. Using UTC for date conversion ensures uniformity
+        # and avoids issues related to different time zones and daylight saving changes.
+        first_commit_date = (
+            datetime.fromtimestamp(first_commit.commit_time, tz=timezone.utc)
+            .date()
+            .isoformat()
+        )
+        most_recent_commit_date = (
+            datetime.fromtimestamp(most_recent_commit.commit_time, tz=timezone.utc)
+            .date()
+            .isoformat()
+        )
+
+        # Prepare the data structure
+        data = {
+            "repo_path": str(repo_path),
+            "total_normalized_entropy": normalized_total_entropy,
+            "number_of_commits": len(commits),
+            "number_of_files": len(file_names),
+            "time_range_of_commits": (first_commit_date, most_recent_commit_date),
+            "file_level_entropy": file_entropy,
+        }
+        return data
+
+    except Exception as e:
+        # If processing fails, return an error dictionary
+        return {"repo_path": str(repo_path), "error": str(e)}
+
+
+def process_repo_for_analysis(
+    repo_url: str,
+) -> Tuple[Optional[float], Optional[str], Optional[str], Optional[int]]:
+    """
+    Processes GitHub repository URL's to calculate entropy and other metadata.
+    This is used to prepare data for analysis, particularly for the seedbank notebook
+    that process PUBMED repositories.
+
+    Args:
+        repo_url (str): The URL of the GitHub repository.
+
+    Returns:
+        tuple: A tuple containing the normalized total entropy, the date of the first commit,
+               the date of the most recent commit, and the total time of existence in days.
+    """
+    temp_dir = tempfile.mkdtemp()
+    try:
+        repo_path = clone_repository(repo_url)
+        # Load the cloned repo
+        repo = pygit2.Repository(str(repo_path))
+
+        # Retrieve the list of commits from the repo
+        commits = get_commits(repo)
+        # Select the first and most recent commits from the list
+        first_commit = commits[-1]
+        most_recent_commit = commits[0]
+
+        # Calculate the time span of existence between the first and most recent commits in days
+        time_of_existence = (
+            most_recent_commit.commit_time - first_commit.commit_time
+        ) // (24 * 3600)
+        # Calculate the time span between commits in days. Using UTC for date conversion ensures uniformity
+        # and avoids issues related to different time zones and daylight saving changes.
+        first_commit_date = (
+            datetime.fromtimestamp(first_commit.commit_time, tz=timezone.utc)
+            .date()
+            .isoformat()
+        )
+        most_recent_commit_date = (
+            datetime.fromtimestamp(most_recent_commit.commit_time, tz=timezone.utc)
+            .date()
+            .isoformat()
```

Comment:
Seeing this pattern pop up again made me wonder about the statements mentioned above. Consider reusability alongside the other statements I mentioned.